### PR TITLE
3.x optimize compactor

### PR DESCRIPTION
**Pull request by davisp**
## Overview

Reviving an old compactor optimization from PR #806.  I'm also going to add in some work from #1006 to give us better visibility on compaction id_seq copy phases.

## Testing recommendations

`$ make check`

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

File: src/couch/src/couch_file.erl [hunk 1]
```diff
@@ -651,6 +726,31 @@ read_raw_iolist_int(#file{fd = Fd, pread_limit = Limit} = F, Pos, Len) ->
...
+            {_Fd, Filepath} = get(couch_file_fd),
+            throw({read_beyond_eof, Filepath});
+        Size when Size > Limit ->
+            couch_stats:increment_counter([pread, exceed_limit]),
+            {_Fd, Filepath} = get(couch_file_fd),
+            throw({exceed_pread_limit, Filepath, Limit});
+        _ ->
+            {Pos, TotalBytes}
+        end
+    end, PosLens),
```
Comment (nickva) [inline src/couch/src/couch_file.erl:746]: Minor nit: could create a separate `get_pread_locnum(Pos, Len) ->` that can be reused with `read_raw_iolist_int` function above.

**Comment by nickva** [comment]
Found a bug when running `./compact_bench.py  -r 50 -n 5000000`

```
 db shards/00000000-ffffffff/cbenchdb.1586802992 died with reason {badarith,[{erlang,div,[100
0,16291.3],[]},{couch_bt_engine_compactor,update_compact_task,1,[{file,"src/couch_bt_engine_compactor.erl"},{line,724}]},{couch_emsort,merge_back_bone,5,[{file,"s
rc/couch_emsort.erl"},{line,278}]},{couch_emsort,decimate,2,[{file,"src/couch_emsort.erl"},{line,265}]},{couch_emsort,merge,2,[{file,"src/couch_emsort.erl"},{line
,208}]},{couch_bt_engine_compactor,sort_meta_data,1,[{file,"src/couch_bt_engine_compactor.erl"},{line,515}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{cou
ch_bt_engine_compactor,start,4,[{file,"src/couch_bt_engine_compactor.erl"},{line,75}]}]}
```

**Comment by davisp** [comment]
Ooops. That `div` bug should be fixed now. Forgot to do integer divisions.

File: src/couch/src/couch_file.erl (hunk 1 above)
Comment (davisp) [reply src/couch/src/couch_file.erl]: Done.

**Pull request merged**